Implemented Feedback workflow for teachers and reviewers

EdwardHealy00 opened the pull request:
- Added email notification when a case study approval action is required for deputies and committee
- Implemented email feedback system for case study submitters
- Implemented Pending Case Studies page with file upload and deletion
- Implemented delete pending case study functionality
- Implemented convert to free case study button
- Corrected case studies route naming mistake

EdwardHealy00 committed:
Added email notification when a case study approval action is required for deputies and committee

EdwardHealy00 committed:
Implemented email feedback system for case study submitters

EdwardHealy00 committed:
Implemented Pending Case Studies page with file upload and deletion

EdwardHealy00 committed:
Implemented delete pending case study functionality

EdwardHealy00 committed:
Implemented convert to free case study button

EdwardHealy00 committed:
Corrected case studies route naming mistake

EdwardHealy00 committed:
Added logic to restrict case study evaluation or modification based on rejections status

Ch0ufleur commented on server/app/services/email.service.ts:
```diff
@@ -77,4 +81,94 @@ export class EmailService {
+        }
+
+        const mailOptions = {
+            from: EMAIL_USERNAME,
+            to: email,
+            subject: `You case study named ${caseStudy.title}` + (isPreApproved ? `has been pre-approved` : `has been declined by a deputy`), 
+            text: `Your case study named ${caseStudy.title}, authored by ${caseStudy.authors}` + 
+                    (isPreApproved ? ` is now ready to get reviewed by the scientific committee.`
+                                   : ` requires changes for its pre-approval. The following criterias were not respected: \n\n` +`${criteriaText}`) +
+                    `\n\n Click here to see its status: http://localhost:3000/my-pending-case-studies`
+
```
On veut pas le texte en français?

EdwardHealy00 replied:
T'a raison. Les autres champs étaient déjà en anglais donc j'ai continué. Mais je vais toute traduire.

Ch0ufleur commented on server/app/controllers/caseStudy.controller.ts:
```diff
@@ -115,23 +258,53 @@ export class CaseStudyController {
+                }
+
+                if (isApproved) {
+                    caseStudy.status += 1; 
```
y'a pas ++ ?

EdwardHealy00 replied:
Oui lol, je vais changé

Ch0ufleur commented on server/app/controllers/caseStudy.controller.ts:
```diff
@@ -88,6 +101,130 @@ export class CaseStudyController {
             }
         });
 
+        this.router.delete('/deleteFile/:filename', async (req: Request, res: Response) => {
+            try {
+                var isSuccessful = await this.caseStudyService.deleteCaseStudyFile(req.params.filename);
+                if (!isSuccessful) {
+                    res.status(404).json('Le fichier ' + req.params.filename+  ' n\'a pas pu être supprimé');
+                    return;
+                }
+                if(!req.body.caseStudy.isRejected) {
+                    res.status(405).json('Il est interdit de modifier une étude de cas en cours d\'évaluation');
+                    return;
+                }
+                
+                res.status(200).json({
+                    status: 'success',
+                });
+            } catch (err: any) {
+                console.log(err);
+            }
+        });
```
pas sur de voir la nuance entre deleteFile et removeFiles plus bas vite de même

EdwardHealy00 replied:
DeleteFile ça supprime le file en tant que telle de la collection, tandis que removeFiles enlève les références au fichiers supprimés du case study. Je vais renommé les routes AddFiles et RemoveFiles pour que ça soit plus clair.

Ch0ufleur commented on server/app/controllers/caseStudy.controller.ts:
```diff
@@ -88,6 +101,130 @@ export class CaseStudyController {
+            }
+        });
+
+        this.router.delete('/delete/:id', async (req: Request, res: Response) => {
+            try {
+                var isSuccessful = await this.caseStudyService.deleteCaseStudy(req.params.id);
+                if (!isSuccessful) {
+                    res.status(404).json('L\'étude de cas n\'a pas pu être supprimé');
+                    return;
+                }
+                
+                res.status(200).json({
+                    status: 'success',
+                });
+            } catch (err: any) {
+                console.log(err);
+            }
+        });
```
est-ce que ces routes là sont restreintes à un utilisateur connecté? je vois que certaines requêtes ont le middlewareRestrictTo activé, il faudrait s'assurer que qqun qui veut delete et qui connait un id peut pas le faire sans les perms appropriées

EdwardHealy00 replied:
Bien vu, j'ai oublié d'ajouter les middlewareRestrictTo

EdwardHealy00 committed:
Fixed PR comments
- translated email service to french
- replaced += 1 with ++
- renamed addFiles and deletesFiles routes
- Restricted patch operations to appropriate roles

Other changes:
- changed convertToFree button to a switch
- Restricted patch operations to be done on rejected cases only

EdwardHealy00 committed:
Removed delete case study functionality
as it was deemed unnecessary

EdwardHealy00 committed:
Added confirmation dialog for modifying study case

EdwardHealy00 committed:
Fixed duplication index bug and table id bug

EdwardHealy00 committed:
Resubmitted case study in comity review status goes back to pre-approval after modification

Ch0ufleur reviewed (approved):
il faudra restreindre la suppresion et modifications de ressources aux users qui les ont créés et aux admins

EdwardHealy00 committed:
Merge branch 'main' into email-feedback
